router/fastdp.go

```diff
@@ -0,0 +1,1115 @@
+	return
+}
+
+func (fwd *fastDatapathForwarder) Stop() {
+	// Might be nice to delete all the relevant flows here, but we
+	// can just let them expire.
+	fwd.fastdp.removeForwarder(fwd.remotePeer.Name, fwd)
+
+	fwd.lock.Lock()
+	defer fwd.lock.Unlock()
+	fwd.sendControlMsg = func([]byte) error { return nil }
+
+	select {
+	case fwd.errorChan <- nil:
+	default:
+	}
+
+	// stop the heartbeat goroutine
+	if !fwd.stopped {
+		fwd.stopped = true
+		close(fwd.stopChan)
+	}
+}
+
+func (fastdp *FastDatapath) addForwarder(peer PeerName,
+	fwd *fastDatapathForwarder) {
+	fastdp.lock.Lock()
+	defer fastdp.lock.Unlock()
+
+	// We shouldn't have two confirmed forwarders to the same
+	// remotePeer, due to the checks in LocalPeer AddConnection.
+	fastdp.forwarders[peer] = fwd
+}
+
+func (fastdp *FastDatapath) removeForwarder(peer PeerName,
+	fwd *fastDatapathForwarder) {
+	fastdp.lock.Lock()
+	defer fastdp.lock.Unlock()
+	if fastdp.forwarders[peer] == fwd {
+		delete(fastdp.forwarders, peer)
+	}
+}
+
+func (fastdp *FastDatapath) deleteFlows() error {
+	fastdp.deleteFlowsCount++
+
+	flows, err := fastdp.dp.EnumerateFlows()
+	if err != nil {
+		return err
+	}
+
+	for _, flow := range flows {
+		err = fastdp.dp.DeleteFlow(flow.FlowKeys)
+		if err != nil && !odp.IsNoSuchFlowError(err) {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (fastdp *FastDatapath) deleteVxlanVports() error {
+	vports, err := fastdp.dp.EnumerateVports()
+	if err != nil {
+		return err
+	}
+
+	for _, vport := range vports {
+		if vport.Spec.TypeName() != "vxlan" {
+			continue
+		}
+
+		err = fastdp.dp.DeleteVport(vport.ID)
+		if err != nil && !odp.IsNoSuchVportError(err) {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (fastdp *FastDatapath) run() {
+	expireMACsCh := time.Tick(10 * time.Minute)
+	expireFlowsCh := time.Tick(5 * time.Minute)
+
+	for {
+		select {
+		case <-expireMACsCh:
+			fastdp.expireMACs()
+
+		case <-expireFlowsCh:
+			fastdp.expireFlows()
+		}
+	}
+}
+
+func (fastdp *FastDatapath) expireMACs() {
+	lock := fastdp.startLock()
+	defer lock.unlock()
+
+	for mac := range fastdp.sendToMAC {
+		if _, present := fastdp.seenMACs[mac]; !present {
+			delete(fastdp.sendToMAC, mac)
+		}
+	}
+
+	fastdp.seenMACs = make(map[MAC]struct{})
+}
+
+func (fastdp *FastDatapath) expireFlows() {
+	lock := fastdp.startLock()
+	defer lock.unlock()
+
+	flows, err := fastdp.dp.EnumerateFlows()
+	if err != nil {
+		log.Warn(err)
+	}
+
+	for _, flow := range flows {
+		if flow.Used == 0 {
+			log.Debug("Expiring flow ", flow.FlowSpec)
+			err = fastdp.dp.DeleteFlow(flow.FlowKeys)
+		} else {
+			fastdp.touchFlow(flow.FlowKeys, &lock)
+			err = fastdp.dp.ClearFlow(flow.FlowKeys)
+		}
+
+		if err != nil && !odp.IsNoSuchFlowError(err) {
+			log.Warn(err)
+		}
+	}
+}
+
+// The router needs to know which flows are active in order to
+// maintain its MAC->peer table.  We do this by quering the router
+// without an actual packet being involed.  Maybe it's
```

Comment:
I think it shows that I was still acclimatising to this laptop's keyboard when I wrote the bulk of fastdp.go.